# Assignment 3: Reproduce Analysis

Skeleton notebook here: 
https://github.com/fedhere/PUI2017_fb55/blob/master/Lab4_fb55/effectivenes%20of%20NYC%20Post-Prison%20Employment%20Programs.ipynb

In [1]:
import os
import sys
import numpy as np
import pylab as pl


import json
import os
# s = json.load( open(os.getenv("PUI2015")+"/fbb_matplotlibrc.json") )
# pl.rcParams.update(s)

%pylab inline

# from IPython.display import Image
# Image(filename='../plotsforclasses/NYCReentryprogram_title.png')

Populating the interactive namespace from numpy and matplotlib


# Notebook example: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05
$H_0: P_0 - P_1 \geq$ 0 <br/>
 $H_a: P_0 - P_1 $< 0 <br/>
 $\alpha$ = 0.05

In [2]:
alpha=0.05

P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

n_0 = 409
n_1 = 564

#lets get the counts by multiplying by the sample size
Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

we must assess the statistical significance


# New parameters: what if we use the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?

** $H_0:$ P_0 - P_1 ≤ 0 ** <br/>
** $H_1:$ P_0 - P_1 > 0 **<br/>
$\alpha$ = 0.05<br/>


Null: the number of prisoners convicted of a felony would be the same or higher in the program that those in the control


In [3]:
alpha=0.05

P2_0 = 11.7 * 0.01
P2_1 = 10 * 0.01

if P2_0 - P2_1  <= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

n2_0 = 409
n2_1 = 568

#lets get the counts by multiplying by the sample size
Nt2_0 = P2_0 * n2_0
Nt2_1 = P2_1 * n2_1


we must assess the statistical significance


# Z Test

### Notebook example

In [4]:
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.421047276465


In [5]:
p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

In [6]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)

20.7697865408


### New parameters: Convicted of a felony

In [7]:
# sample proportion
sp2 = (P2_0 * n2_0 + P2_1 * n2_1) / (n2_1 + n2_0)
print (sp)

0.421047276465


In [8]:
# calculate the standard error
# and to calculate p

p2 = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se2 = lambda p2, n0, n1: np.sqrt(p2 * (1 - p2) * (1.0 / n0 + 1.0 / n1))

In [9]:
z = lambda p0, p1, s : (p0 - p1) / s
z_2y = z(P2_1, P2_0, se2(p2(P2_0, P2_1, n2_0, n2_1), n2_0, n2_1))
print (z_2y)

-0.84764020522


In [10]:
# check the chart to find value and subtract from 1
p_val = 1 - .7995

def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p_val < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p_val < a  else 'not rejected') )

    
report_result(p_val, alpha)

is the p value 0.20 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


___

# The $\chi^2$ test¶


In [11]:
from scipy.stats import chisquare
def evalChisq(values):
    '''Evaluates the chi sq from a contingency value
    Arguments:
    values: 2x2 array or list, the contingengy table
    '''
    if not (len(values.shape) == 2 and values.shape == (2,2)):
        print ("must pass a 2D array")
        return -1
    values = np.array(values)
    E = np.empty_like(values)
    for j in range(len(values[0])):
        for i in range(2):
            
            E[i][j] = ((values[i,:].sum() * values[:,j].sum()) / 
                        (values).sum())
    return ((values - E)**2 / E).sum()


### Notebook example

In [12]:
# recreating the skeleton test
Ntot = 973 # a + b + c + d = tot
sample_values_example = np.array([[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]])
sample_values_example

array([[ 395.364 ,  168.636 ],
       [  12.4745,  394.685 ]])

In [13]:
evalChisq(sample_values_example)

436.22346257536003

### Convicted of Felony

In [14]:
# sample values for "convictions of a felony"
Ntot2 = 977 
sample_values_felony = np.array([[0.1 * 568, 0.9 * 568], [0.117 * 409, 0.883 * 409]])
sample_values_felony

array([[  56.8  ,  511.2  ],
       [  47.853,  361.147]])

In [15]:
evalChisq(sample_values_felony)

0.71849391750528868

In [16]:
chisquare(sample_values_felony, axis=None)

Power_divergenceResult(statistic=649.48411552917082, pvalue=1.8845503190687889e-140)


## What the null means: 
Since there's one degree of freedom, we can look at the chart and compare our findings with 3.85 (since our alpha is .05). Since .718 is less than 3.85 we do not reject the null. We couldn't prove that this couldn't have been due to chance. 



## Fill out the contingency table below

|Convicted of a felony| yes  |  no |
|---|---|---|
|test sample  | 56.8  | 511.2  |
|control sample   |  47.853 | 361.147  |
|total  |104.653| 872.447  |